CASSJAVA-89 Fix deprecated table configs in Cassandra 5

mkars10 opened the pull request:
This fixes CASSJAVA-89 by switching from deprecated `chunk_length_kb` to `chunk_length_in_kb`

Commit:
CASSJAVA-89 fix: support schema options that changed in Cassandra 5.0

Patch by Michael Karsten; reviewed by Abe Ratnofsky and Andy Tolbert for CASSJAVA-89

tolbertam commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -58,6 +58,20 @@ default SelfT withCDC(boolean enabled) {
+  /**
+   * Defines the crc check chance.
+   *
+   * <p>Note that using this option with a version of Apache Cassandra less than 5.0 will raise a
```
It looks like this was added in 3.0: https://github.com/apache/cassandra/blob/115c5539840354d6af81a246d391071d2851f636/NEWS.txt#L1472-L1473 as part of CASSANDRA-9839

```suggestion
   * <p>Note that using this option with a version of Apache Cassandra less than 3.0 will raise a
```

tolbertam commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -58,6 +58,20 @@ default SelfT withCDC(boolean enabled) {
+   */
+  @NonNull
+  @CheckReturnValue
+  default SelfT withCRCCheckChance(double crc_check_chance) {
+    return withOption("crc_check_chance", crc_check_chance);
```
```suggestion
  default SelfT withCRCCheckChance(double crcCheckChance) {
    return withOption("crc_check_chance", crcCheckChance);
```

tolbertam commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -97,22 +111,32 @@ default SelfT withCompaction(@NonNull CompactionStrategy<?> compactionStrategy)
+   * @deprecated This method only exists for backward compatibility. Use {@link
+   *     #withLZ4Compression(int)} instead.
    */
+  @Deprecated
```
:+1: seems reasonable to me as it maintains backwards compatibility with older versions, should also note that this will not work with C* 5.0+

tolbertam commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -121,22 +145,55 @@ default SelfT withLZ4Compression() {
+  @NonNull
+  @CheckReturnValue
+  default SelfT withZstdCompression(int chunkLengthKB) {
+    return withCompression("org.apache.cassandra.io.compress.ZstdCompressor", chunkLengthKB);
```
I think `ZstdCompressor` itself works doesn't it?  Would be nice to use that just for consistency of how we declare the other compressors.

mkars10 replied:
I agree it would be nice, but I don't see the class like I do the [LZ4 compressor](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/internal/core/protocol/Lz4Compressor.java). I didn't want to additionally add the compressor with this PR. Am I overthinking things?

tolbertam replied:
> I agree it would be nice, but I don't see the class like I do the [LZ4 compressor](https://github.com/apache/cassandra-java-driver/blob/4.x/core/src/main/java/com/datastax/oss/driver/internal/core/protocol/Lz4Compressor.java). I didn't want to additionally add the compressor with this PR. Am I overthinking things?

ah, so this refers to the class in the apache source:  https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/io/compress/ZstdCompressor.java

Anything with the package name `org.apache.cassandra.io.compress` can be specified without the full package name:

https://github.com/apache/cassandra/blob/78d6892dfb617ea7e8feee44b2e6fca34cba2780/src/java/org/apache/cassandra/schema/CompressionParams.java#L255

The LZ4 compressor you linked is for the on the wire network compression between a client and a cassandra node, for which Zstd is not supported yet, but that could be nice to support at some point in the future.

mkars10 replied:
Ah, thanks for adding the context I was missing. Just switched over to `ZstdCompressor`

tolbertam commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -97,22 +111,32 @@ default SelfT withCompaction(@NonNull CompactionStrategy<?> compactionStrategy)
+   */
+  @NonNull
+  @CheckReturnValue
+  default SelfT withLZ4Compression(int chunkLengthKB) {
```
One thing I noticed is that both LZ4Compressor and ZStdCompressor support additional options (https://cassandra.apache.org/doc/5.0/cassandra/managing/operating/compression.html#configuring-compression) which we can't currently configure here as far as I can tell.   It might be nice to find some way of passing these through, but maybe we don't need to solve this on this PR (unless you are motivated to do so), maybe we can do that as a follow on?

mkars10 replied:
I guess the easiest way would be to provide an option similar to [withCompression](https://github.com/apache/cassandra-java-driver/blob/bb9bb11e573258e12e4272379f88e4d8b5f103a0/query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java#L183) that also takes a map of additional options, but that raises the question of if/how to do input validation. 

Technically, users could do this already by using `withOption(String, Object)` directly, so I'd rather do this as a follow on since this fixes a potential block to 5.0 upgrades as-is.

tolbertam replied:
agreed, sounds good to defer that as a follow on 👍

tolbertam reviewed (commented):
This is looking pretty good to me, I had some suggestions, willing to follow up with another quick review based on your thoughts!

tolbertam commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -185,19 +252,30 @@ default SelfT withCompression(@NonNull String compressionAlgorithmName) {
+  default SelfT withCompression(@NonNull String compressionAlgorithmName, int chunkLengthKB) {
+    return withOption(
+        "compression",
+        ImmutableMap.of("class", compressionAlgorithmName, "chunk_length_in_kb", chunkLengthKB));
```
Does feel a bit weird to use `chunk_length_in_kb` for this implementation, but `chunk_length_kb` for the deprecated methods, but this feels like the best way to maintain backwards compatibility for now.  When we can rid of the deprecated methods, `chunk_length_kb` goes away :+1:

mkars10 commented:
@tolbertam thanks for the review! I addressed most of your points. Lmk about the Zstd compressor, if I'm being dense and there's a better way, I'm all for it.

tolbertam commented:
> @tolbertam thanks for the review! I addressed most of your points. Lmk about the Zstd compressor, if I'm being dense and there's a better way, I'm all for it.

Changes looking good to me.  I'm out of town until Monday, but I'll work to find a second reviewer and do some testing with it when I get back.

aratno commented on query-builder/src/test/java/com/datastax/oss/driver/api/querybuilder/schema/CreateTableTest.java:
Would you be open to adding an integration test to check query-builder paths? It would be good to get a test failure on future syntax changes like this one, rather than discovering the issue later when attempting to startup an application on the newer Cassandra version. You could follow `com.datastax.oss.driver.querybuilder.JsonInsertIT` as an example.

mkars10 replied:
That's a good idea. It looks like I need to do some more local setup to get integration tests running though. I can take a look at that in a few days.

aratno replied:
Docs here: https://github.com/apache/cassandra-java-driver/blob/4.x/CONTRIBUTING.md#integration-tests

Let me know if you have any questions - happy to help you get set up.

mkars10 replied:
@aratno I added 3 integration tests to cover the following cases.
- top-level `crc_check_chance` for C* >= 3.0
- `chunk_length_in_kb` for C* >= 5.0
- `chunk_length_kb` for C* < 5.0

aratno commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -185,19 +252,30 @@ default SelfT withCompression(@NonNull String compressionAlgorithmName) {
+
+  /**
+   * @deprecated This method only exists for backward compatibility. Will not work with Apache
+   *     Cassandra 5.0 or later. Use {@link #withCompression(String, int)} instead.
```
For these docstrings that indicate version support, could you update to include a reference to the relevant Jira?
```suggestion
   *     Cassandra 5.0 or later due to removal of deprecated table properties (<a href="https://issues.apache.org/jira/browse/CASSANDRA-18742">CASSANDRA-18742</a>). Use {@link #withCompression(String, int)} instead.
```

aratno commented on query-builder/src/main/java/com/datastax/oss/driver/api/querybuilder/schema/RelationOptions.java:
```diff
@@ -58,6 +58,20 @@ default SelfT withCDC(boolean enabled) {
+   * <p>Note that using this option with a version of Apache Cassandra less than 3.0 will raise a
+   * syntax error.
+   *
+   * <p>If no call is made to this method, the default value set is 0.99.
```
Looks like the default is 1.0, not 0.99. The default is decided on the server and is documented here: https://cassandra.apache.org/doc/latest/cassandra/reference/cql-commands/create-table.html

You can remove this part of the docstring.

tolbertam reviewed (approved):
:+1: great work, and thank you for your persistence and patience on working through this!
